### PR TITLE
Fix  add enabled in _initRecordAnnotations()

### LINKED ISSUE #153571: torch.cuda.memory._record_memory_history(enabled=None) does not clean up previously added hooks
### 🐛 Describe the bug

It seems like torch.cuda.memory._record_memory_history(enabled=None) doesn't clear previously added hooks (and in fact it adds hooks unconditionally). The relevant code is here:

https://github.com/pytorch/pytorch/blob/5e6e52e7c9e29338e6b011e9bb1eeaab9fb6e2e4/torch/csrc/cuda/memory_snapshot.cpp#L173

Here `_initRecordAnnotations()` is called unconditionally and it adds hooks:

https://github.com/pytorch/pytorch/blob/c92ea3bc98b0bd601d1e899e76c40286999a6133/torch/csrc/cuda/memory_snapshot.cpp#L100-L118

I don't see a removing of annotations anywhere in the code.

Here is a small benchmark that shows the regression:

```
import torch
import time

def alloc_benchmark():
  start = time.time()
  for _ in range(100000):
    a = torch.empty((2**1,), device="cuda")
    b = torch.empty((2**5,), device="cuda")
    c = torch.empty((2**10,), device="cuda")
    d = torch.empty((2**20,), device="cuda")
  end = time.time()
  return end - start

warmup_duration = alloc_benchmark()

baseline_duration = alloc_benchmark()
baseline_duration2 = alloc_benchmark()

torch.cuda.memory._record_memory_history(enabled=None, stacks="python")

new_duration = alloc_benchmark()

print(f"baseline duration: {baseline_duration}")
print(f"baseline duration2: {baseline_duration2}")
print(f"new duration: {new_duration}")
pct_diff2 = (baseline_duration2 - baseline_duration) / baseline_duration * 100
pct_diff = (new_duration - baseline_duration) / baseline_duration * 100
print(f"new vs. baseline diff: {pct_diff}")
print(f"baseline vs. baseline diff: {pct_diff2} (for sanity checking noise)")

```

This will print something like this:

```
baseline duration: 0.7567446231842041
baseline duration2: 0.757319450378418
new duration: 0.7591335773468018
new vs. baseline diff: 0.31568829026435585
baseline vs. baseline diff: 0.07596052573127365 (for sanity checking noise)
```

Although the diff is small, performance sensitive users may care about this.

cc: @LucasLLC @vishal9-team 

### Versions

python collect_env.py 
Collecting environment information...
PyTorch version: 2.7.0a0+gitaceef9f
Is debug build: False
CUDA used to build PyTorch: 12.0
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-5)
Clang version: 19.1.7 (CentOS 19.1.7-1.el9)
CMake version: version 3.31.4
Libc version: glibc-2.34

Python version: 3.12.8 | packaged by Anaconda, Inc. | (main, Dec 11 2024, 16:31:09) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.4.3-0_fbk20_zion_2830_g3e5ab162667d-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H100
GPU 1: NVIDIA H100
GPU 2: NVIDIA H100
GPU 3: NVIDIA H100
GPU 4: NVIDIA H100
GPU 5: NVIDIA H100
GPU 6: NVIDIA H100
GPU 7: NVIDIA H100

Nvidia driver version: 535.154.05
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.8.9.2
/usr/lib64/libcudnn.so.9.2.0
/usr/lib64/libcudnn_adv.so.9.2.0
/usr/lib64/libcudnn_adv_infer.so.8.9.2
/usr/lib64/libcudnn_adv_train.so.8.9.2
/usr/lib64/libcudnn_cnn.so.9.2.0
/usr/lib64/libcudnn_cnn_infer.so.8.9.2
/usr/lib64/libcudnn_cnn_train.so.8.9.2
/usr/lib64/libcudnn_engines_precompiled.so.9.2.0
/usr/lib64/libcudnn_engines_runtime_compiled.so.9.2.0
/usr/lib64/libcudnn_graph.so.9.2.0
/usr/lib64/libcudnn_heuristic.so.9.2.0
/usr/lib64/libcudnn_ops.so.9.2.0
/usr/lib64/libcudnn_ops_infer.so.8.9.2
/usr/lib64/libcudnn_ops_train.so.8.9.2
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      52 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             384
On-line CPU(s) list:                0-383
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 9654 96-Core Processor
CPU family:                         25
Model:                              17
Thread(s) per core:                 2
Core(s) per socket:                 96
Socket(s):                          2
Stepping:                           1
Frequency boost:                    enabled
CPU(s) scaling MHz:                 76%
CPU max MHz:                        3707.8120
CPU min MHz:                        1500.0000
BogoMIPS:                           4792.82
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid overflow_recov succor smca fsrm flush_l1d
Virtualization:                     AMD-V
L1d cache:                          6 MiB (192 instances)
L1i cache:                          6 MiB (192 instances)
L2 cache:                           192 MiB (192 instances)
L3 cache:                           768 MiB (24 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-95,192-287
NUMA node1 CPU(s):                  96-191,288-383
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Vulnerable: eIBRS with unprivileged eBPF
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] flake8==6.1.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.15.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.14.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.1.0
[pip3] optree==0.13.0
[pip3] torch==2.7.0a0+gitaceef9f
[conda] numpy                     2.1.0                    pypi_0    pypi
[conda] optree                    0.13.0                   pypi_0    pypi
[conda] torch                     2.7.0a0+gitaceef9f           dev_0    <develop>
[conda] torchfix                  0.4.0                    pypi_0    pypi

cc @ptrblck @msaroufim @eqy @jerryzh168 @robieta @chaekit @guotuofeng @guyang3532 @dzhulgakov @davidberard98 @briancoutinho @sraikund16 @sanrise

### PR DESCRIPTION
**Pull request by ghostspiders**
Fixes #153571
torch.cuda.memory._record_memory_history(enabled=None) does not clean up previously added hooks.Add (enabled=None) check in _initRecordAnnotations()

cc @ptrblck @msaroufim @eqy @jerryzh168

### CODE REVIEW

**Comment by ghostspiders** [comment]
> ## 🔗 Helpful Links
> ### 🧪 See artifacts and rendered test results at [hud.pytorch.org/pr/153794](https://hud.pytorch.org/pr/153794)
> * 📄 Preview [Python docs built from this PR](https://docs-preview.pytorch.org/pytorch/pytorch/153794/index.html)
> * 📄 Preview [C++ docs built from this PR](https://docs-preview.pytorch.org/pytorch/pytorch/153794/cppdocs/index.html)
> * ❓ Need help or want to give feedback on the CI? Visit the [bot commands wiki](https://github.com/pytorch/pytorch/wiki/Bot-commands) or our [office hours](https://github.com/pytorch/pytorch/wiki/Dev-Infra-Office-Hours)
> 
> Note: Links to docs will display an error until the docs builds have been completed.
> 
> ## ❌ 1 New Failure
> As of commit [3a0fb9f](https://github.com/pytorch/pytorch/commit/3a0fb9f32bf51c7099a2c7154309bec03fb929bc) with merge base [6803419](https://github.com/pytorch/pytorch/commit/68034198e5c25f10005d13c6a68cb7a4e3a76f1e) (image):
> 
> NEW FAILURE - The following job has failed:
> 
> * [Check Labels / Check labels](https://hud.pytorch.org/pr/pytorch/pytorch/153794#42432910127) ([gh](https://github.com/pytorch/pytorch/actions/runs/15096963064/job/42432910127))
> 
> This comment was automatically generated by Dr. CI and updates every 15 minutes.

module: cuda

**Review by cyyever — APPROVED** [review]
